packages/synthetics-sdk-broken-links/src/broken_links.ts

```diff
@@ -108,7 +130,62 @@ export async function runBrokenLinks(
     followed_links
   );
 }
+/* c8 ignore stop */
+
+/**
+ * Retrieves all links on the page using Puppeteer, handling relative and
+ * protocol-relative links and filtering for HTTP/HTTPS links.
+ *
+ * @param page - The Puppeteer page instance to retrieve the links from.
+ * @param query_selector_all - The CSS selector to identify link elements.
+ * @param get_attributes - An array of attribute names to retrieve from the link elements.
+ * @returns An array of LinkIntermediate objects representing the links found.
+ */
+export async function retrieveLinksFromPage(
+  page: Page,
+  query_selector_all: string,
+  get_attributes: string[]
+): Promise<LinkIntermediate[]> {
+  const origin_url = await page.url();
+  // `page.evaluate()` runs in a different context, isolated from main Node.js,
+  // so c8 is not able to track coverage.
+  /* c8 ignore start */
+  return await page.evaluate(
+    (
+      query_selector_all: string,
+      get_attributes: string[],
+      origin_url: string
+    ) => {
+      const link_elements: HTMLElement[] = Array.from(
+        document.querySelectorAll(query_selector_all)
+      );
+      return link_elements.flatMap((link_element: HTMLElement) => {
+        const anchor_text = link_element?.textContent?.trim() ?? '';
+
+        return get_attributes
+          .map((attr) => (link_element.getAttribute(attr) || '').toString())
+          .filter((value) => {
+            const qualifed_url = new URL(value, origin_url);
+            return value && qualifed_url.href.startsWith('http');
+          })
+          .map((value) => {
+            const qualifed_url = new URL(value, origin_url);
```

Comment:
I'm going to ask a dumb question, where does URL come from? I thought that you would have needed to require it somewhere (https://nodejs.org/api/url.html#url)

Reply:
I'm honestly not 100% sure but this link I think suggests it's not necessary to import it...
https://nodejs.org/api/url.html#the-whatwg-url-api